src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,690 @@
+
+        Returns:
+            message which was received
+
+    ***************************************************************************/
+
+    public ReceivedMessage receive ( T ... ) ( MessageFiber fiber, T messages )
+    {
+        foreach ( message; messages )
+            static assert(is(typeof(message) == Message));
+
+        auto event = this.nextEvent(fiber, messages);
+        enforce(event.active == event.active.message,
+            "Unexpected event: waiting only for message receipt");
+        return event.message;
+    }
+
+    /***************************************************************************
+
+        Unregisters all waiting events of `fiber` and kills the fiber.
+
+        Params:
+            fiber = fiber to unregister and abort
+
+    ***************************************************************************/
+
+    public void abort ( MessageFiber fiber )
+    {
+        this.waiting_fibers.length = Array.moveToEnd(this.waiting_fibers,
+            WaitingFiber(fiber),
+            ( Const!(WaitingFiber) e1, Const!(WaitingFiber) e2 )
+            {
+                return e1.fiber == e2.fiber;
+            }
+        );
+
+        fiber.kill();
+    }
+
+    /***************************************************************************
+
+        Triggers a signal with the specified code. Any fiber which is awaiting
+        this signal will be resumed in the next eventLoop cycle.
+
+        Params:
+            conn = request-on-conn event dispatcher (required in order to
+                resume the request-on-conn fiber, if it is currently suspended)
+            code = signal code to trigger
+
+    ***************************************************************************/
+
+    public void signal ( RequestOnConnBase.EventDispatcher conn, ubyte code )
+    {
+        if ( !conn.resumeFiber(code) )
+            this.queued_signals ~= code;
+    }
+
+    /***************************************************************************
+
+        Runs the event loop cycle (wait for events, notify waiting fibers) until
+        no waiting fibers are registered.
+
+        Note that the event loop can be aborted by aborting each fiber which has
+        events registered (see abort()).
+
+        Params:
+            conn = request-on-conn event dispatcher to multiplex fiber I/O
+            events over
+
+    ***************************************************************************/
+
+    public void eventLoop ( RequestOnConnBase.EventDispatcher conn )
+    {
+        while ( this.waiting_fibers.length )
+        {
+            Const!(void)[] received;
+
+            if ( this.waitingWriters() )
+            {
+                auto writer = this.popWaitingWriter();
+                bool sent;
+
+                do
+                {
+                    this.dispatchQueuedSignals(conn);
+
+                    auto resume_code = conn.sendReceiveAndHandleEvents(
```

Comment:
Updated with handling for request-on-conn exceptions. (All waiting fibers are resumed with the exception.)